### PR TITLE
Added a new projection generation for congruence

### PR DESCRIPTION
**Pull request by StamesJames**
The new projection generation can now generate projections for some dependently typed fields. For this it looks if the type can be expressed as an applicative term consisting only of terms present in the environment or that can be extracted from parameters or indices of the inductive type that is getting projected. One issue I still have is with universes. Even though the generation isn't able to handle match terms right now it should be possible when the exact match statement is extracteable from an inductive index. When I try this however it doesn't work because of something with the universes of the terms.

Fixes / closes #????

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [ ] Added / updated **documentation**.
  
  - [ ] Documented any new / changed **user messages**.
  - [ ] Updated **documented syntax** by running `make doc_gram_rsts`.

- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by StamesJames** [comment]
Running make test-suite and the linter worked on my computer. The pipeline complains that the Seq.find_map function isn't known. Should I only use the Coq specific collections like CArray and CList? But I think there is no CSeq right?

**Comment by mrhaandi** [comment]
> The pipeline complains that the Seq.find_map function isn't known.

Coq is supposed to be compatible with `"ocaml" {>= "4.09.0"}`, `Seq.find_map` is since `ocaml 4.14` in [Seq](https://ocaml.org/manual/5.2/api/Seq.html).

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by mrhaandi** [comment]
`ci-metacoq` shows an actual regression. The following code works prior to the current PR:
```
Record R (A : Type) (B : A -> Type) : Type := RI { a : A; b : B a }.

Goal forall A B (x : A) (y1 y2 : B),
  y1 <> y2 -> RI _ _ x y1 <> RI _ _ x y2.
Proof.
  intros.
  congruence.
```

**Comment by StamesJames** [comment]
I think I fixed this now. I am currently running the CI and will see if it is fixed.

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
I don't think that the debugging brings much, it's adding a lot of noise to the code and if ever we need some printf debugging we can already implement it locally when considering a specific snipped. I'd recommend removing it.

Also I'm not a huge fan of the rendering of the code based on parentheses. This makes it hard to follow the control flow. Most of it seems due to the inclusion of imperative calls in the middle of otherwise pure code. If you remove the debug code a good part of it should go away, but for what remains I prefer using the idiom
```ocaml
let () = effectful code in
next
```
to
```ocaml
effectful code;
next
```
The previous style also prevents having to wrap this into parentheses.

**Comment by ppedrot** [comment]
I'm not a huge fan of leaving dead code around, but YMMV. Note that I also personally mostly use the OCaml debugger for this kind of stuff, and the datatypes you want to print probably don't need a dedicated printer in this case anyway.

File: plugins/cc/ccprojectability.ml [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+  in pr_type_composition_rec type_composition level
+
+type projection_type =
+| Simple
+| Dependent of type_composition
+| NotProjectable
+
+(*find a composition to form a given term by extracting from given terms*)
+let find_type_composition
+  env sigma (cnstr:Names.constructor * EConstr.EInstance.t) (argindex:int) (env_field_type:EConstr.t) (ind:Names.inductive EConstr.puniverses) (env_ind_params:EConstr.t array) (env_ind_args:EConstr.t array)
```
Comment (ppedrot) [inline plugins/cc/ccprojectability.ml:113]: Nit: explicitly typing the arguments is quite heavyweight and not idiomatic.

File: plugins/cc/ccprojectability.ml [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+  let (rel_field_context, (rel_field_annot, rel_field_type), rel_field_rest) = get_ith_arg sigma (argindex+env_ind_n_params) rel_type_of_constructor in
+  let (rel_target_context, rel_target_type) = EConstr.decompose_prod sigma rel_field_rest in
+  let rel_field_type_lifted = EConstr.Vars.lift (List.length rel_target_context + 1) rel_field_type in
+  let (_,rel_args) = EConstr.decompose_app sigma rel_target_type in
+  let (rel_ind_params, rel_ind_args) = CArray.chop env_ind_n_params rel_args in
+  let rec find_type_composition_rec env sigma (rel_term_to_compose:EConstr.t) (env_term_to_compose:EConstr.t)=
+    let rel_term_to_compose_kind = EConstr.kind sigma rel_term_to_compose in
+    match rel_term_to_compose_kind with
+    | Var _ | Const _ | Ind _ | Construct _ -> (sigma, Some (FromEnv rel_term_to_compose))
+    | _ ->(
```
Comment (ppedrot) [inline plugins/cc/ccprojectability.ml:126]: You don't need to wrap the contents of a pattern-matching branch in parentheses (and neither of if-then-else). In case you need nested matches (and only when needed), to work around the deficient OCaml syntax I'd recommend wrapping the inner match in a begin-end block, like:
```ocaml
match foo with
| Some x ->
  begin match bar with
  | None -> ...
  | Some y -> ...
  end
| None -> ...
```

**Comment by SkySkimmer** [comment]
In addition to @ppedrot comments:
Test suite needs fixing (I guess because of the stdlib split)
A doc update or changelog entry is also needed

**Comment by StamesJames** [comment]
> In addition to @ppedrot comments: Test suite needs fixing (I guess because of the stdlib split) A doc update or changelog entry is also needed

Yes, there are tests where I use the Std Vectors. I will just put the definition of vectors in there.
I am also working on some refactoring to make the code a bit easier to read. I will push it in the next days

**Comment by StamesJames** [comment]
I removed the dependency of the Coq stdlib in my test cases and refactored my code a bit, to hopefully make it more readable.

**Comment by mrhaandi** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@StamesJames this still needs a changelog (run dev/tools/make-changelog.sh, follow its instructions, commit & push the result)

**Comment by StamesJames** [comment]
> @StamesJames this still needs a changelog (run dev/tools/make-changelog.sh, follow its instructions, commit & push the result)

Sorry totally lost track of this. I pushed a changelog and rebased to the current state of Rocq now.

File: doc/changelog/04-tactics/19700-dependent-projectability-congruence-Changed.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+- **Changed:**
+  the congruence tactic now can handle some dependently typed constructor fields.
+  The field's type has to be composed of terms occuring globally or projectable from
```
Comment (SkySkimmer) [inline doc/changelog/04-tactics/19700-dependent-projectability-congruence-Changed.rst:3]: linter is failng due to this
```suggestion
  The field's type has to be composed of terms occuring globally or projectable from
```

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Comment by ppedrot** [comment]
I will write a follow-up PR to make the style more in line with the rest of the Rocq repo, it feels currently a bit alien compared to the standard style.
